326 bake config sparse

kporangehat opened the pull request:
Adds a --sparse option that will skip caching app_store bundles to keep baked configs smaller. 

Fixes a bug in tk-core that would skip checking if an engine's apps needed to be cached if the parent engine was skipped.

kporangehat committed:
for #326: add option for baking sparse configs

Adds a --sparse option that will skip caching app_store bundles to keep baked configs smaller.

kporangehat committed:
for #326: fix for not checking caching of apps if the engine skipped.

Ensures that when an engine is flagged to be skipped in the caching process, the apps in the engine are still checked.
Previously, the engine's apps were not checked when the engine was skipped.

kporangehat committed:
for #326: remove unused imports

gplsteph commented on developer/bake_config.py:
```diff
@@ -1,12 +1,8 @@
-# Copyright (c) 2013 Shotgun Software Inc.
+# Copyright 2018 GPL Solutions, LLC.  All rights reserved.
```
In this case we want to preserve the original headers and push back these changes onto the SG repo. I have a PR for it on Shotgun Software. So please put back the Shotgun Software header.

gplsteph commented on developer/bake_config.py:
```diff
@@ -93,8 +102,10 @@ def bake_config(sg_connection, config_uri, target_path, do_zip=False):
+    :param config_uri: A TK config descriptor uri.
     :param target_path: Path to build
+    :param do_zip: Optionally zip up config once it's baked. Defaults to False.
+    :param make_sparse: Don't cache app_store bundles into the config. Defaults to False.
```
What about naming this param `sparse_caching`? `make_sparse` is a bit vague to me?

gplsteph commented on developer/bake_config.py:
```diff
@@ -50,6 +42,23 @@
 BUNDLE_CACHE_ROOT_FOLDER_NAME = "bundle_cache"
 
 
+def _studio_should_skip_caching(desc):
```
Having _studio_ in this method name does not seem right, as we intent to submit these changes for a PR on SG repo. What about `should_skip_caching_sparse`?

gplsteph commented on developer/bake_config.py:
```diff
@@ -127,22 +138,18 @@ def bake_config(sg_connection, config_uri, target_path, do_zip=False):
-    cache_apps(sg_connection, config_descriptor, bundle_cache_root)
 
-    # Now analyze what core the config needs and cache it if needed.
-    if config_descriptor.associated_core_descriptor:
```
Just checking: from now on we never cache tk-core, even if there is an associated core descriptor? Should we cache it if the sparse option is off? What if we have a special tk-core release we want to pull from our github and not from the app store?

gplsteph commented on developer/bake_config.py:
```diff
@@ -127,22 +138,18 @@ def bake_config(sg_connection, config_uri, target_path, do_zip=False):
-            bundle_cache_root_override=bundle_cache_root
-        )
-        associated_core_desc.ensure_local()
+        caching._should_skip_caching = _studio_should_skip_caching
```
That does not seem right. Why not adding an extra param to `cache_apps` which would be a callable to call when deciding what to skip or not? The default value could either be `_should_skip_caching`, or `None` if preferred. For the latter case, the `cache_apps` code would fallback to calling `_should_skip_caching` if the param is `None`, and blindly call the callable otherwise? This would give some flexibility to decide what needs to be cached or not and would solve a general case, as opposed to just our own very specific case?

kporangehat replied:
Do we need this to be a callable or could we just pass in the descriptor types we want to skip as a `list` of `str`s (for example `should_skip=["dev", "path", "app_store"]`). I agree callable would be more flexible but as it stands we'd only have the descriptor type as a `str` to work with so there's not a ton we could do with that.

gplsteph commented on developer/utils/caching.py:
```diff
@@ -81,15 +81,14 @@ def cache_apps(sg_connection, cfg_descriptor, bundle_cache_root):
-                desc,
-                bundle_cache_root
-            )
+            if not _should_skip_caching(desc):
```
See my previous comments, but I feel the callable should be a param of `cache_apps`, and fall back by default on `_should_skip_caching`.

gplsteph reviewed (changes requested):
Requesting changes for the following reasons:
- I think we still want to cache tk-core if sparse caching is off. If it is on, we shouldn't cache it only if dealing with an app store release.
- Please revisit the `_should_skip_caching` override. I feel the right approach would be to give the callable as a param.
- Our goal is to submit these changes back onto SG tk-core, so headers should be preserved and we shouldn't have any "studio" naming.
Thanks!

kporangehat committed:
for #326: revert copyright notice change

kporangehat committed:
for #326: add callable param to cache_apps to check whether to cache

Adds an additional parameter to cache_apps() which will be used to check the descriptor type and determine whether to cache the bundle or not. Defaults to using the internal _should_skip_caching() function if not provided.

kporangehat committed:
for #326: Ensure we are caching tk-core if needed.

- some additional comment/docstring updates

kporangehat commented:
Ready for review again.

gplsteph commented on developer/bake_config.py:
```diff
@@ -30,15 +28,13 @@
 
 from utils import (
-    cache_apps, authenticate, add_authentication_options,
+    caching, cache_apps, authenticate, add_authentication_options,
```
Do we need to import `caching` now that we don't monkey patch it anymore?

gplsteph commented on developer/bake_config.py:
```diff
@@ -127,22 +141,41 @@ def bake_config(sg_connection, config_uri, target_path, do_zip=False):
-    if config_descriptor.associated_core_descriptor:
-        logger.info("Config is specifying a custom core in config/core/core_api.yml.")
-        logger.info("This will be used when the config is executing.")
+    # If sparse_caching is True, we add app_store descriptor types to the skip
```
```
# If spare_caching is True we use our own descriptor filter which skips app_store descriptor 
#to keep our bundle cache small and let Toolkit download the bundles from the Apps store
```
?
Saying _we add app_store descriptors_ is a bit confusing here...

gplsteph commented on developer/bake_config.py:
```diff
@@ -127,22 +141,41 @@ def bake_config(sg_connection, config_uri, target_path, do_zip=False):
+    # list in order to keep our bundle cache small and let Desktop handle
+    # downloading app_store bundles separately.
+    should_skip_caching_callable = None
+    if sparse_caching:
```
I think something like:
```
if sparse_caching:
   logger.info("Performing sparse caching. Will not cache standard app_store bundles")
    cache_apps(sg_connection, config_descriptor, bundle_cache_root, _should_skip_caching_sparse)
else:
    cache_apps(sg_connection, config_descriptor, bundle_cache_root)
```
would be a lot clearer.

gplsteph commented on developer/bake_config.py:
```diff
@@ -127,22 +141,41 @@ def bake_config(sg_connection, config_uri, target_path, do_zip=False):
+    if core_descriptor:
+        logger.info("Config defines a specific core in config/core/core_api.yml.")
+        logger.info("This will be used when the config is executing.")
+        if not sparse_caching or not _should_skip_caching_sparse(core_descriptor):
```
A comment would be helpful here, something like `If sparse_caching is on, check if we need to cache tk-core or not`.

gplsteph reviewed (approved):
Looks good to me, a couple of small things could be possibly improved, or not.
Thanks.

kporangehat committed:
for #326: addressing CR comments
